packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -1829,77 +1834,20 @@ namespace Ifpack2 {
-            for (local_ordinal_type i=0;i<blocksize;++i)
-              packed_multivector(pri, i, col)[vi] = scalar_multivector(blocksize*lclrow(ri0+j)+i,col);
-        } else if (TagType::id > 0) { //ToScalarMultiVector
-          const impl_scalar_type df = damping_factor;
-          for (local_ordinal_type col=0;col<num_vectors;++col) 
-            for (local_ordinal_type i=0;i<blocksize;++i) {
-              impl_scalar_type &y = scalar_multivector(blocksize*lclrow(ri0+j)+i,col);
-              const impl_scalar_type yc = packed_multivector(pri, i, col)[vi];
-              if (TagType::id == 1) y  = df*yc;
-              else                  y += df*(yc - y); 
-            }
-        }
-      }
-
-      template<typename TagType>
-      KOKKOS_INLINE_FUNCTION
-      void copy_multivectors_with_norm(const local_ordinal_type &j, 
-                                       const local_ordinal_type &vi, 
-                                       const local_ordinal_type &pri, 
-                                       const local_ordinal_type &/* nrow */,  
-                                       const local_ordinal_type &ri0,
-                                       /* */ magnitude_type *norm) const {
-        if (TagType::id > 0) { //ToScalarMultiVector
-          const impl_scalar_type df = damping_factor;
-          for (local_ordinal_type col=0;col<num_vectors;++col) {
-            const local_ordinal_type offset = col*blocksize;
-            for (local_ordinal_type i=0;i<blocksize;++i) {
-              impl_scalar_type &y = scalar_multivector(blocksize*lclrow(ri0+j)+i,col);
-              const impl_scalar_type yc = packed_multivector(pri, i, col)[vi];
-              const impl_scalar_type yd = TagType::id == 1 ? yc : yc - y;
-              const magnitude_type abs_yd = Kokkos::ArithTraits<impl_scalar_type>::abs(yd);
-              norm[offset+i] += abs_yd*abs_yd;
-              if (TagType::id == 1) y  = df*yc;
-              else                  y += df*yd; 
-            }
-          }
-        }
+	for (local_ordinal_type col=0;col<num_vectors;++col) 
+	  for (local_ordinal_type i=0;i<blocksize;++i)
+	    packed_multivector(pri, i, col)[vi] = scalar_multivector(blocksize*lclrow(ri0+j)+i,col);
       }
 
     public:
 
-      // local reduction of norms with runtime array 
-      // this value type and value_count is required for Kokkos
-      typedef magnitude_type value_type[];
-      int value_count;
-
-      KOKKOS_INLINE_FUNCTION
-      void 
-      init(magnitude_type *dst) const {
-        for (int i=0;i<value_count;++i)         
-          dst[i] = Kokkos::reduction_identity<magnitude_type>::sum();
-      }
-
-      KOKKOS_INLINE_FUNCTION
-      void 
-      join(volatile magnitude_type *dst, const volatile magnitude_type *src) const {
-        for (int i=0;i<value_count;++i) 
-          dst[i] += src[i];
-      }          
-
       MultiVectorConverter() = delete;
       MultiVectorConverter(const MultiVectorConverter &b) = default;
```

Comment:
Is there any particular reason you `default` the copy constructor, but not the assignment operator?  If you don't actually need this, it's better to leave it off, since it suppresses implicit declaration of a move constructor and move assignment operator.

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-five

If the default operations are fine here, you could just go with the Rule of Zero: 

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-zero

Reply:
Good point. I just delete it because the class always need part interface. Probably when I create the class, I just want to prevent accidental mistake myself. I will just delete the deleted constructor. I will fix.